xls/modules/dbe/lz4_encoder.x

```diff
@@ -0,0 +1,832 @@
+    let fifo = update(fifo, u32:0, data);
+    fifo
+}
+
+fn RamZeroReq<ADDR_WIDTH: u32, DATA_WIDTH: u32, NUM_PARTITIONS: u32>
+    () -> RamReq<ADDR_WIDTH, DATA_WIDTH, NUM_PARTITIONS> {
+    RamReq<ADDR_WIDTH, DATA_WIDTH, NUM_PARTITIONS> {
+        addr:uN[ADDR_WIDTH]:0,
+        data:uN[DATA_WIDTH]:0,
+        write_mask: (),
+        read_mask: (),
+        we: false,
+        re: false,
+    }
+}
+
+fn RamZeroResp<DATA_WIDTH: u32>
+    () -> RamResp<DATA_WIDTH> {
+    RamResp<DATA_WIDTH> {
+        data:uN[DATA_WIDTH]:0,
+    }
+}
+
+fn RamWriteReq<NUM_PARTITIONS: u32, ADDR_WIDTH: u32, DATA_WIDTH: u32>
+    (addr:uN[ADDR_WIDTH], data:uN[DATA_WIDTH])
+    -> RamReq<ADDR_WIDTH, DATA_WIDTH, NUM_PARTITIONS> {
+    RamReq {
+        addr:addr,
+        data:data,
+        write_mask: (),
+        read_mask: (),
+        we: true,
+        re: false,
+    }
+}
+
+fn RamReadReq<DATA_WIDTH: u32, NUM_PARTITIONS: u32, ADDR_WIDTH: u32>
+    (addr:uN[ADDR_WIDTH])
+    -> RamReq<ADDR_WIDTH, DATA_WIDTH, NUM_PARTITIONS> {
+    RamReq {
+        addr:addr,
+        data:uN[DATA_WIDTH]:0,
+        write_mask: (),
+        read_mask: (),
+        we: false,
+        re: true,
+    }
+}
+
+enum FsmSt: u4 {
+    RESET = 0,
+    HASH_TABLE_CLEAR = 1,
+    RESTART = 2,
+    FIFO_PREFILL = 3,
+    FIFO_POSTFILL = 4,
+    START_MATCH_0 = 5,
+    START_MATCH_1 = 6,
+    CONTINUE_MATCH_0 = 7,
+    CONTINUE_MATCH_1 = 8,
+    EMIT_SHORT_MATCH = 9,
+    EMIT_FINAL_LITERALS = 10,
+    EMIT_END = 11,
+    ERROR = 12,
+}
+
+struct State<SYM_WIDTH: u32, PTR_WIDTH: u32, CNT_WIDTH: u32, HASH_BITS: u32,
+                MINMATCH: u32, FINAL_LITERALS: u32, FIFO_CACHE_SZ: u32,
+                FIFO_IN_SZ: u32> {
+    fifo_cache: uN[SYM_WIDTH][FIFO_CACHE_SZ],
+    fifo_in: uN[SYM_WIDTH][FIFO_IN_SZ],
+    fifo_in_count: u32,
+    fifo_in_nvalid: u32,
+    wp: uN[PTR_WIDTH],
+    hb_all_valid: bool,
+    pnew: uN[PTR_WIDTH],
+    pold: uN[PTR_WIDTH],
+    match_nconts: u32,
+    ht_ptr: uN[HASH_BITS],
+    recycle: bool,
+    finalize: bool,
+    fsm: FsmSt,
+}
+
+fn init_state<SYM_WIDTH: u32, PTR_WIDTH: u32, CNT_WIDTH: u32, HASH_BITS: u32,
+                MINMATCH: u32, FINAL_LITERALS: u32, FIFO_CACHE_SZ: u32,
+                FIFO_IN_SZ: u32>()
+                -> State<SYM_WIDTH, PTR_WIDTH, CNT_WIDTH, HASH_BITS,
+                            MINMATCH, FINAL_LITERALS, FIFO_CACHE_SZ,
+                            FIFO_IN_SZ> {
+    State{
+        fifo_cache: uN[SYM_WIDTH][FIFO_CACHE_SZ]: [uN[SYM_WIDTH]:0, ...],
+        fifo_in: uN[SYM_WIDTH][FIFO_IN_SZ]: [uN[SYM_WIDTH]:0, ...],
+        fifo_in_count: u32:0,
+        fifo_in_nvalid: u32:0,
+        wp: uN[PTR_WIDTH]:0,
+        hb_all_valid: false,
+        pnew: uN[PTR_WIDTH]:0,
+        pold: uN[PTR_WIDTH]:0,
+        match_nconts: u32:0,
+        ht_ptr: uN[HASH_BITS]:0,
+        recycle: false,
+        finalize: false,
+        fsm: FsmSt::RESET
+    }
+}
+
+pub proc encoder_base<
```

Comment:
Add a comment on how matches are found using the hash table (i.e. algorithm).
Comment on how aliases in hashing are found and prevented from causing a bad encoding.

Reply:
I'm going to prepare a document that explains how the algorithm works - this will be explained there.

Reply:
@rdob-ant curious which format do you plan to use? (separate markdown file, comments in the header of the impl, or google docs?)

Reply:
@proppy I wrote a plain markdown file.